clusterloader2/pkg/measurement/common/pod_command.go

```diff
@@ -0,0 +1,697 @@
+			Stdout: stdoutBuf,
+			Stderr: stderrBuf,
+		})
+		execErrChan <- err
+	}()
+
+	t := time.NewTimer(params.Timeout)
+	defer t.Stop()
+
+	// Two different cases: (1) if the command returns before the timeout, and (2) if the timeout
+	// triggers before the command is done.
+	// If the timeout triggers, then the command isn't cancelled. This logic isn't available until
+	// client-go version 0.26 (see Executor.StreamWithContext).
+	select {
+	case err = <-execErrChan:
+		endTime = time.Now()
+		if err == nil {
+			break
+		}
+
+		switch e := err.(type) {
+		case exec.CodeExitError:
+			p.statsLock.Lock()
+			p.stats.NonZeroRCs++
+			p.statsLock.Unlock()
+
+			exitCode = e.ExitStatus()
+
+			klog.V(2).Infof(
+				"%s: warning: non-zero exit code %d for named command %s in pod %s/%s",
+				podPeriodicCommandMeasurementName, exitCode, params.Name, pod.Namespace, pod.Name,
+			)
+		default:
+			return nil, err
+		}
+	case <-t.C:
+		p.statsLock.Lock()
+		p.stats.Timeouts++
+		p.statsLock.Unlock()
+
+		endTime = time.Now()
+		hitTimeout = true
+
+		klog.V(2).Infof(
+			"%s: warning: hit timeout of %s for named command %s in pod %s/%s",
+			podPeriodicCommandMeasurementName, params.Timeout.String(), params.Name, pod.Namespace, pod.Name,
+		)
+	}
+
+	klog.V(4).Infof(
+		"%s: finished running named command %s in pod %s/%s",
+		podPeriodicCommandMeasurementName, params.Name, pod.Namespace, pod.Name,
+	)
+
+	return &runCommandResult{
+		stdout:     stdoutBuf.String(),
+		stderr:     stderrBuf.String(),
+		ExitCode:   exitCode,
+		Name:       params.Name,
+		Command:    params.Command,
+		Timeout:    params.Timeout.String(),
+		HitTimeout: hitTimeout,
+		StartTime:  startTime,
+		EndTime:    endTime,
+	}, nil
+}
+
+// runAllCommandsInPod runs all of the configured commands in the given specific pod.
+func (p *podPeriodicCommandMeasurement) runAllCommandsInPod(pod *v1.Pod) {
+	wg := &sync.WaitGroup{}
+	resultCh := make(chan *runCommandResult, len(p.params.Commands))
+
+	getRunCommandFunc := func(c *podPeriodicCommandMeasurementCommandParams) func() {
+		return func() {
+			defer wg.Done()
+
+			r, err := p.runCommandInPod(pod, c)
+			if err != nil {
+				p.statsLock.Lock()
+				p.stats.Errors++
+				p.statsLock.Unlock()
+
+				klog.V(2).Infof(
+					"%s: error while running named command %s in %s-%s: %v",
+					podPeriodicCommandMeasurementName, c.Name, pod.Namespace, pod.Name, err.Error(),
+				)
+
+				p.errOnce.Do(func() {
+					p.err = err
+				})
+
+				return
+			}
+
+			if r != nil {
+				resultCh <- r
+			}
+		}
+	}
+
+	klog.V(4).Infof(
+		"%s: running commands on pod %s/%s", podPeriodicCommandMeasurementName, pod.Namespace, pod.Name,
+	)
+
+	for _, command := range p.params.Commands {
+		wg.Add(1)
+		go getRunCommandFunc(command)()
+	}
+
+	wg.Wait()
+	close(resultCh)
+
+	klog.V(4).Infof(
+		"%s: finished running commands on pod %s/%s", podPeriodicCommandMeasurementName, pod.Namespace, pod.Name,
+	)
+
+	if len(resultCh) == 0 {
```

Comment:
I am wondering if it would make sense to skip this check and also plumb `err` from `runCommandInPod` to `runCommandResult`. That would give us a few improvements:
- we would see that command were executed in results
- we would get all errors + we wouldn't need to use `errOnce`
- we could return `runAllCommandsResult` in this method and append it in `commandWorker`.
What do you think?

Reply:
This makes sense! I finished making these changes and will test them tomorrow. I also added some more parameters to help control when the measurement should fail, which I think could be useful, ie, if the measurement should fail on non-zero RC, if a timeout was hit, or a client-go related error